Feat/ Multi limit pools throttler on Ascend Ex

Nullably opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Create a new Multi limit pool throttler class to handle rate limit where a call can increase multiple pools (e.g. Ascend Ex)
- Create based on VariedRateAPIThrottler, but could not use the same base class as functions signatures differ
- Update Ascend Ex exchange to use this new throttle class, also update the `_api_request` fn to better capture and report errors.
- The Ascend Ex rate limits taken from https://www.notion.so/hummingbot/Ascendex-Rate-Limits-8e60e4bab9a14fa0b55926f9c5a601e6 
- The functionality of this new throttler encompasses VariedRateAPIThrottler and FixedRateAPIThrottler, ie. you can use this instead of those 2.

**Tests performed by developer**
- Create and run detailed unit tests on new API throttler class.
- Test AscendEx on PMM by manually lower rate limits, the strategy works, the throttler applies delay as needed and warning message logged when the limit is about to reached.

**Tips to QA**
- The json error reported on #3754 log file obscures actual error returned from the exchange so we don't know what the actual errors were, this PR solves this, so please look out for new error messages.

aarmoa commented:
We need to create a separate task to refactor the API Throttlers and produce a unified model. The current implementation is good to solve the issue, but will make it difficult for developers to reuse and support in the future.

Nullably commented:
@aarmoa Agree, I'll create another ticket for this.

aarmoa reviewed (dismissed):
I discussed the implementation with @Nullably and this is useful to solve the current issue with the limits in Ascendex. We will create a separate image to refactor the solution and unify the API Throttler implementation with the Multi limit pools.

RC-13 commented:
Based on the initial testing, there are random instances that the API error 300011 occurs within one of the test bots however, it didn't cause any inconsistent order creation nor stuck/failed to cancel orders. For the liquidity mining strategy, the occurrence of API error 300011 is more common which I think is okay since there are multiple markets dividing on the asset. Currently running 3 bots for overnight testing and check back the result before approving the PR.

rxlxrxsx commented:
When running the connector using liquidity mining strategy and when the rate limit warning starts to show, on random instances I see that it shows orders that were already cancelled from earlier showing on `status`, particularly orders with order age more than 10 (order refresh time is currently set 10 to with order refresh tolerance disabled). There's also one order that shows an order age of n/a.

<img width="1423" alt="Screen Shot 2021-07-21 at 2 51 19 PM" src="https://user-images.githubusercontent.com/64377055/126454409-6ebe8321-bb8f-4d3b-8c00-936c7e9b124a.png">

Logs and config: [ascendex_lm.zip](https://github.com/CoinAlpha/hummingbot/files/6853909/ascendex_lm.zip)

RC-13 commented:
After running some test bots overnight, the API rate limit was hit and cause one order to get stuck or uncancelled while the other bots got an HTTP 429 error which wasn't handled properly as it filled the status window with HTML logs:
![http429](https://user-images.githubusercontent.com/57189990/126571380-57574252-4ec1-432d-9d65-edb87b694876.png)
![xdai](https://user-images.githubusercontent.com/57189990/126571395-57b6fe72-15d3-4a45-87e7-8d0bdbec3571.png)
[testbot logs.zip](https://github.com/CoinAlpha/hummingbot/files/6859246/testbot.logs.zip)

Nullably commented:
@RC-13 There are 3 log files in the zip file, are they all from the same trading account? and running from bots from same IP address?

RC-13 commented:
@Nullably Yes, same account and same IP

Nullably commented:
> @Nullably Yes, same account and same IP

The bot can't handle rate limits sharing yet, it assumes it has total 100% share of the limits.

But this is good to test to see how the bot handles in case of rate limit exceeded.

RC-13 reviewed (approved):
LGTM

Nullably commented:
@RC-13 As I think as it is common that users run multi bots on the same account, I am adding a new feature for those bots to share the rate limits, will update the PR again soon. I am going to fix how the HTML error was shown as well.

Nullably commented:
I added  a new global config `rate_limits_share_pct`, so that a portion of rate limit can be assigned to each bot as needed in case where multiple bots are running from the same account.
![image](https://user-images.githubusercontent.com/37262506/126609206-1a6b5780-eedf-4676-85b7-9c4efd7004e1.png)
I updated the warning message to notify user as well when the rate is about to reach, the limit here is a the rate limit * rate_limits_share_pct.
![image](https://user-images.githubusercontent.com/37262506/126609665-7395c748-d4d2-4829-80fc-2b707fe79737.png)

Nullably commented:
@RC-13 Can you please retest this? and try the new global config if it works well with multiple bots.
@aarmoa Can you please re-review this?
Thanks

PtrckM commented:
- though its configurable its not showing on `config` command 
![image](https://user-images.githubusercontent.com/54516858/126724843-b1e1da5c-3b47-46d4-a750-2b8ca3adf6ab.png)

Nullably commented:
> * though its configurable its not showing on `config` command

just fixed.

aarmoa reviewed (approved):
Changes look good to me
